java-checks/src/main/java/org/sonar/java/checks/PrivateFieldUsedLocallyCheck.java

```diff
@@ -0,0 +1,285 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.java.cfg.CFG;
+import org.sonar.java.cfg.CFG.Block;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Symbol.TypeSymbol;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S1450")
+public class PrivateFieldUsedLocallyCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE = "Remove the \"%s\" field and declare it as a local variable in the relevant methods.";
+
+  private Map<MethodTree, Set<Symbol>> fieldsLiveInMethodEntryPerMethod;
+  private Set<Symbol> fieldsReadOnAnotherInstance;
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (tree.is(Kind.CLASS)) {
+      TypeSymbol classSymbol = ((ClassTree) tree).symbol();
+      initMethodsLiveFields(classSymbol);
+
+      for (Symbol memberSymbol : classSymbol.memberSymbols()) {
+
+        if (isPrivateField(memberSymbol)
+          && !memberSymbol.usages().isEmpty()
+          && !isUsedOutsideMethods(memberSymbol, classSymbol)) {
+
+          checkPrivateField(memberSymbol);
+        }
+      }
+    }
+  }
+
+  private boolean isUsedOutsideMethods(Symbol privateFieldSymbol, TypeSymbol classSymbol) {
+    for (IdentifierTree usageIdentifier : privateFieldSymbol.usages()) {
+      Tree containingClassOrMethod = containingClassOrMethod(usageIdentifier);
+
+      if (containingClassOrMethod.is(Kind.CLASS)
+        || containingClassOrMethod.is(Kind.METHOD) && !((MethodTree) containingClassOrMethod).symbol().owner().equals(classSymbol)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS864)